packages/replay-experimental/src/suspense/reactInternalsCaches.ts

```diff
@@ -0,0 +1,331 @@
+          const functionOutline = findFunctionOutlineForLocation(location, symbols);
+
+          if (functionOutline) {
+            functionDetails = {
+              functionName,
+              functionOutline,
+              lineIndex: lineZeroIndex + 1,
+            };
+          }
+        }
+      }
+    }
+
+    return functionDetails;
+  },
+});
+
+// TODO This only handles a single ReactDOM source file.
+const reactDomLinesCache: Cache<
+  [
+    replayClient: ReplayClientInterface,
+    functionName: ReactInternalFunctionName,
+    reactDomSource: SourceDetails
+  ],
+  string[]
+> = createSingleEntryCache({
+  debugLabel: "ReactDomLines",
+  async load([replayClient, functionName, reactDomSource]) {
+    const streaming = streamingSourceContentsCache.stream(replayClient, reactDomSource.id);
+    await streaming.resolver;
+
+    const reactDomSourceContents = streaming.value!;
+    return reactDomSourceContents.split("\n");
+  },
+});
+
+export const reactInternalMethodsHitsIntervalCache = createFocusIntervalCacheForExecutionPoints<
+  [replayClient: ReplayClientInterface, functionName: ReactInternalFunctionName],
+  PointDescription
+>({
+  debugLabel: "ReactInternalMethodHits",
+  getPointForValue: data => data.point,
+  async load(rangeStart, rangeEnd, replayClient, functionName) {
+    if (!rangeStart || !rangeEnd) {
+      return [];
+    }
+
+    const functionDetails = await reactInternalMethodsDetailsCache.readAsync(
+      replayClient,
+      functionName
+    );
+
+    if (!functionDetails) {
+      return [];
+    }
+
+    const reactDomSources = await sourcesByPartialUrlCache.readAsync(replayClient, "react-dom.");
+
+    if (!reactDomSources.length) {
+      return [];
+    }
+
+    // TODO There could be more than one here. We'll take the first.
+    const reactDomSource = reactDomSources[0];
+
+    // We found a function outline, so we can use that to get the hits.
+    const hitPoints = await hitPointsCache.readAsync(
+      BigInt(rangeStart),
+      BigInt(rangeEnd),
+      replayClient,
+      { ...functionDetails.functionOutline.breakpointLocation!, sourceId: reactDomSource.id },
+      null
+    );
+
+    return hitPoints;
+  },
+});
+
+export type ReactUpdateScheduled = {
+  type: "scheduled";
+  cause: "user" | "internal" | "unknown";
+} & FormattedPointStack;
+
+export type ReactSyncUpdatedStarted = {
+  type: "sync_started";
+} & FormattedPointStack;
+
+export type ReactRenderCommitted = {
+  type: "render_committed";
+} & FormattedPointStack;
+
+export type ReactRenderEntry =
+  | ReactUpdateScheduled
+  | ReactSyncUpdatedStarted
+  | ReactRenderCommitted;
+
+export const reactRendersIntervalCache = createFocusIntervalCacheForExecutionPoints<
+  [replayClient: ReplayClientInterface],
+  ReactRenderEntry
+>({
+  debugLabel: "ReactRenders",
+  getPointForValue: data => data.point.point,
+  async load(rangeStart, rangeEnd, replayClient) {
```

Comment:
This function has a lot of serial async operations. Looks like the first one is blocking, but the last few could be run in parallel.